src/mozanalysis/metrics/__init__.py

```diff
@@ -51,6 +59,28 @@ def _check_experiments_column_type(self, attribute, value):
                 f"{repr(self.EXPERIMENT_COLUMN_TYPES)}"
             )
 
+    @default_dataset.validator
+    def _check_default_dataset_provided_if_needed(self, attribute, value):
+        self.from_expr_for(None)
+
+    def from_expr_for(self, dataset: Optional[str]) -> str:
+        """Expands the ``from_expr`` template for the given dataset.
+        If ``from_expr`` is not a template, returns ``from_expr``.
+
+        Args:
+            dataset (str or None): Dataset name to substitute
+                into the from expression.
+        """
+        effective_dataset = dataset or self.default_dataset
```

Comment:
Do we _need_ to be this friendly with the error messages? If not, `.format()` has got behaviour pretty close to what we need?
```
kwargs = {'dataset': effective_dataset} if effective_dataset else {}
return self._from_expr.format(**kwargs)
```

Reply:
I was tempted but I don't think we'll regret the verbosity even if it adds a few loc.